### PR TITLE
Curl response refactoring

### PR DESCRIPTION
**Pull request by nupplaphil**
FollowUp #5862 

- extended Curl to parse Curl Response
- refactored Network::curl()
- replaced every Network::curl() execution with the new Curl container
- refactored Network::post()
- replaced every Network::post() execution with the new Curl container
- refactored Network::getCurl()
- replaced every Network::getCurl() execution with a Curl Response

### CODE REVIEW

**Comment by nupplaphil** [comment]
- tests locally succeed
- node https://friendica.philipp.info is up and running
=> pages work
=> posted a test to myself on https://opensocial.at (which appears)
=> daemon is working as well (logs shows no exception output)

File: src/Network/Curl.php [hunk 1]
```diff
@@ -3,61 +3,207 @@
 namespace Friendica\Network;
 
 
+use Friendica\Network\HTTPException\InternalServerErrorException;
+
 /**
  * A content class for Curl call results
  */
 class Curl
```
Comment (MrPetovan) [inline src/Network/Curl.php]: Please rename the class `CurlResult` since it doesn't perform any curl calls.

File: src/Network/CurlResult.php [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+	/**
+	 * Creates an errored CURL response
+	 *
+	 * @param string $url optional URL
+	 *
+	 * @return CurlResult a CURL with error response
+	 */
+	public static function createErrorCurl($url = '')
+	{
+		return new CurlResult($url,'', [ 'http_code' => 0 ]);
```
Comment (MrPetovan) [inline src/Network/CurlResult.php]: Code standards: Please add a space after commas and remove spaces inside square brackets.

**Comment by nupplaphil** [comment]
Todo
[ ] headers aren't fully parsed
[ ] adding a CurlResponseTest

**Comment by MrPetovan** [comment]
Is it a problem for this PR?

**Comment by nupplaphil** [comment]
I'm not sure. I think my test is faulty not the code ^^ .. Have to check it

**Comment by MrPetovan** [comment]
I mean, are the non-fully parsed headers a liability, or is it just a nice-to-have?

**Comment by nupplaphil** [comment]
OK I made a debug logging output and on my node, all headers are here. It's just a testing problem. So no, it shouldn't be a problem.

**Comment by annando** [comment]
Just out of interest: What is the advantage of not directly returning an array with data but returning an object that need to be called with some other function to fetch the data that is already stored in that object but cannot be accessed directly?

**Comment by nupplaphil** [comment]
Tests (mocking is possible, explicit testing of CurlResult) and Encapsulation (the user doesn't have to know about the inner structure of a CurlResult because it shows the possibilities)

**Comment by nupplaphil** [comment]
and we don't have to check the same parts of code again and again (f.e. what is a timeout, when was the call a success)

**Comment by nupplaphil** [comment]
And there was already a mix of array usage and get-methods ( $a->getCurlCode )

**Comment by annando** [comment]
I definitely prefer arrays, since they are transparently returning all of their content and they can be used without any additional effort.

This "getCurlCode" had been some very ugly relict of the past (and of course needed to be replaced), so I think it is okay that it is gone now.

If this had been needed for improving the tests, then it's okay for me. But we really, really, really shouldn't replace all place where we work with arrays with this "get" and "set" functions, just because we can or just because there **might** be some situation in the future where this **might** be an advantage.

**Comment by MrPetovan** [comment]
Using arrays everywhere directly led to #5470, so there's a liability. Objects ensure that your data structure is consistent, which prevent most type mismatch and missing index errors. It isn't always warranted, but if you're regularly checking if an array contains the expected keys, then it might be useful to switch.

**Comment by annando** [comment]
Objects are a PITA when you are using JSON and your elements have names like "@id" or "as:to" (and you have to check for the existence of some object elements).

But especially I don't really like the idea of reading and writing every object variable with a special getter or setter function. Abstraction is a good thing, but I prefer abstraction with sure instinct.

**Comment by MrPetovan** [comment]
I don't like getters and setters as well in general. I find them redundant with public properties.

**Comment by nupplaphil** [comment]
They are just redundant if you allow to write properties at every time.
So for REAL public properties, where we want to override, it's a good idea. But in general, I'm more to restrict accessing variables.

=> It makes the code more harden and testable.
=> I don't think it's a good pattern to allow writing values in every case.

What do you think about this OOP-Pattern:
http://www.php.net/manual/en/language.oop5.overloading.php ?

**Comment by MrPetovan** [comment]
I used to find PHP class magic methods awesome, but it makes the code harder to read for humans and machine alike, so I stopped using them.

**Comment by nupplaphil** [comment]
@annando in my belief, patterns are not just about a religion. It's more like a "best practice" (especially in teams with different developers). But you're right, we shall not be too dogmatic about it.

F.e. for my thoghts the OOP-pattern is maybe hard to implement, but makes the code a lot more easier to maintain. There are a lot more benefits which will directly improve the code quality in Friendica. f.e. the flexibility of changing arrays and public properties every time makes it hard to refactor or adopt code. Because (from my perspective as a "newbie Friendica code") I have to mention every occurrence of the properties in the code when I want to change parts of it. That's the strength of encapsulation were the business logic of a part of code occur once.

I know it's maybe a philosophic discussion, so I think I will 'just' add code and you will probable say if I make things too complicated :-)

**Comment by annando** [comment]
AFAIK you are coming from an OOP world with C++ and so on? I'm earning my money with some language that isn't OOP at all. I mostly don't understand this whole stuff at all. When you say " is maybe hard to implement" I guess it will make it very hard for me to understand and totally impossible to work with it for many of our part time contributors who aren't professional developers at all.

I know that - for example - static classes are disliked by many professionals. But I see them as a good compromise. You can encapsulated local functions and variables while on the other hand you can access them like regular functions.

I, personally, will always prefer the static classes.

**Pull request merged**